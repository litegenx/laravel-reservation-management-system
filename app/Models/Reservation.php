<?php
declare(strict_types=1);

namespace App\Models;

use App\Helpers\Traits\TimeHelper;
use Doctrine\DBAL\Schema\Column;
use Eloquent;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasOne;
use Illuminate\Foundation\Http\FormRequest;
use Illuminate\Support\Carbon;
use Staudenmeir\EloquentEagerLimit\HasEagerLimit;
use Technote\CrudHelper\Models\Contracts\Crudable as CrudableContract;
use Technote\CrudHelper\Models\Traits\Crudable;
use Technote\SearchHelper\Models\Contracts\Searchable as SearchableContract;
use Technote\SearchHelper\Models\Traits\Searchable;

/**
 * App\Models\Reservation
 *
 * @property int $id
 * @property int $guest_id 利用者ID
 * @property int $room_id 部屋ID
 * @property Carbon $start_date 利用開始日
 * @property Carbon $end_date 利用終了日
 * @property string $checkout チェックアウト時間
 * @property string $status ステータス
 * @property Carbon|null $created_at
 * @property Carbon|null $updated_at
 * @method static Builder|Reservation newModelQuery()
 * @method static Builder|Reservation newQuery()
 * @method static Builder|Reservation query()
 * @method static Builder|Reservation whereCreatedAt($value)
 * @method static Builder|Reservation whereEndDate($value)
 * @method static Builder|Reservation whereGuestId($value)
 * @method static Builder|Reservation whereId($value)
 * @method static Builder|Reservation whereNumber($value)
 * @method static Builder|Reservation whereRoomId($value)
 * @method static Builder|Reservation whereStartDate($value)
 * @method static Builder|Reservation whereUpdatedAt($value)
 * @method static Builder|Reservation whereCheckout($value)
 * @method static Builder|Reservation whereStatus($value)
 * @property-read bool $is_future
 * @property-read bool $is_past
 * @property-read bool $is_present
 * @property-read Carbon $start_datetime
 * @property-read Carbon $end_datetime
 * @property-read string $end_date_str
 * @property-read string $start_date_str
 * @property-read ReservationDetail $detail
 * @property-read int $charge
 * @property-read int $days
 * @property-read int $nights
 * @property-read string $stays
 * @property-read Guest $guest
 * @property-read Room $room
 * @property-read int $payment
 * @mixin Eloquent
 * @mixin Builder
 */
class Reservation extends Model implements CrudableContract, SearchableContract
{
    use HasEagerLimit, TimeHelper, Crudable, Searchable;

    /**
     * @var array
     */
    protected $guarded = [
        'id',
    ];

    /**
     * @var array
     */
    protected $casts = [
        'guest_id'   => 'int',
        'room_id'    => 'int',
        'start_date' => 'date',
        'end_date'   => 'date',
    ];

    /**
     * @var array
     */
    protected $appends = [
        'start_date_str',
        'end_date_str',
        'start_datetime',
        'end_datetime',
        'is_past',
        'is_present',
        'is_future',
        'days',
        'nights',
        'stays',
        'charge',
    ];

    /**
     * @var array
     */
    protected $with = [
        'detail',
        'room',
        'room.reservations',
    ];

    /**
     * @var int
     */
    protected $perPage = 10;

    protected static function boot()
    {
        parent::boot();

        self::updating(function ($model) {
            /** @var Reservation $model */
            if ($model->isDirty('guest_id')) {
                $model->detail->copyGuestData(Guest::find($model->guest_id));
            }
            if ($model->isDirty('room_id')) {
                $model->detail->copyRoomData(Room::find($model->room_id));
            }
            $model->detail->save();
        });
    }

    /**
     * @return array
     */
    public static function getSearchRules(): array
    {
        return [
            'start_date' => 'filled|date',
            'end_date'   => 'filled|date',
            'room_id'    => 'filled|integer|exists:rooms,id',
        ];
    }

    /**
     * @return array
     */
    public static function getSearchAttributes(): array
    {
        return [
            'start_date' => __('request.reservations.start_date'),
            'end_date'   => __('request.reservations.end_date'),
            'room_id'    => __('request.reservations.room_id'),
        ];
    }

    /**
     * @param  Builder  $query
     * @param  array  $conditions
     */
    protected static function setConditions(Builder $query, array $conditions)
    {
        if (! empty($conditions['s'])) {
            collect($conditions['s'])->each(function ($search) use ($query) {
                $query->where(function ($builder) use ($search) {
                    /** @var Builder $builder */
                    $builder->where('reservation_details.guest_name', 'like', "%{$search}%")
                            ->orWhere('reservation_details.guest_name_kana', 'like', "%{$search}%")
                            ->orWhere('reservation_details.guest_address', 'like', "%{$search}%")
                            ->orWhere('reservation_details.room_name', 'like', "%{$search}%");
                });
            });
        }

        if (! empty($conditions['start_date'])) {
            $query->whereDate('reservations.end_date', '>=', $conditions['start_date']);
        }
        if (! empty($conditions['end_date'])) {
            $query->whereDate('reservations.start_date', '<', $conditions['end_date']);
        }

        if (! empty($conditions['room_id'])) {
            $query->where('reservations.room_id', $conditions['room_id']);
        }
    }

    /**
     * @return array
     */
    protected static function getSearchJoins(): array
    {
        return [
            'reservation_details' => [
                'first'  => 'reservation_details.reservation_id',
                'second' => 'reservations.id',
            ],
        ];
    }

    /**
     * @return array
     */
    protected static function getSearchOrderBy(): array
    {
        return [
            'reservations.start_date' => 'desc',
            'reservations.id'         => 'desc',
        ];
    }

    /**
     * @return array
     */
    public static function getCrudListRelations(): array
    {
        return [
            'guest',
        ];
    }


    /**
     * @return array
     */
    public static function getCrudDetailRelations(): array
    {
        return [
            'guest',
        ];
    }

    /**
     * @return array
     */
    public static function getCrudUpdateRelations(): array
    {
        return [
            'detail' => ReservationDetail::class,
        ];
    }

    /**
     * @param  array  $rules
     * @param  string  $name
     * @param  Column  $column
     * @param  bool  $isUpdate
     * @param  int|null  $primaryId
     * @param  FormRequest  $request
     *
     * @return array
     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
     */
    public static function filterCrudRules(array $rules, string $name, Column $column, bool $isUpdate, ?int $primaryId, FormRequest $request): array
    {
        if ('reservations.end_date' === $name) {
            $rules[] = 'after_or_equal:reservations.start_date';
        }
        if ('reservation_details.number' === $name) {
            $rules['min'] = 'min:1';

            if ($isUpdate) {
                $roomId = request()->input('reservations.room_id', static::findOrFail($primaryId)->room_id);
            } else {
                $roomId = request()->input('reservations.room_id');
            }
            if ($roomId) {
                $room = Room::find($roomId);
                if ($room) {
                    $rules['max'] = 'max:'.$room->number;
                }
            }
        }
        if (in_array($name, [
            'reservation_details.number',
            'reservation_details.room_name',
            'reservation_details.guest_name',
            'reservation_details.guest_name_kana',
            'reservation_details.guest_zip_code',
            'reservation_details.guest_address',
            'reservation_details.guest_phone',
        ])) {
            unset($rules['filled']);
            unset($rules['required']);
            $rules['nullable'] = 'nullable';
        }

        return self::addReservationRule($rules, $name, $request);
    }

    /**
     * @param  array  $rules
     * @param  string  $name
     * @param  FormRequest  $request
     *
     * @return array
     */
    private static function addReservationRule(array $rules, string $name, FormRequest $request)
    {
        if ($request->has('reservations.start_date')) {
            if ('reservations.start_date' === $name) {
                $rules[] = 'reservation_term';
                $rules[] = 'reservation_availability';
                $rules[] = 'reservation_duplicate';
            }
        } elseif ($request->has('reservations.end_date')) {
            if ('reservations.end_date' === $name) {
                $rules[] = 'reservation_term';
                $rules[] = 'reservation_availability';
                $rules[] = 'reservation_duplicate';
            }
        } elseif ($request->has('reservations.room_id')) {
            if ('reservations.room_id' === $name) {
                $rules[] = 'reservation_availability';
            }
        } elseif ($request->has('reservations.guest_id')) {
            if ('reservations.guest_id' === $name) {
                $rules[] = 'reservation_duplicate';
            }
        }

        return $rules;
    }

    /**
     * @return HasOne
     */
    public function detail(): HasOne
    {
        return $this->hasOne(ReservationDetail::class);
    }

    /**
     * @return BelongsTo
     */
    public function guest(): BelongsTo
    {
        return $this->belongsTo(Guest::class)->without('reservations');
    }

    /**
     * @return BelongsTo
     */
    public function room(): BelongsTo
    {
        return $this->belongsTo(Room::class)->without('reservations');
    }

    /**
     * @return string
     */
    public function getStartDateStrAttribute(): string
    {
        return $this->start_date->format('Y-m-d');
    }

    /**
     * @return string
     */
    public function getEndDateStrAttribute(): string
    {
        return $this->end_date->format('Y-m-d');
    }

    /**
     * @return Carbon
     */
    public function getStartDatetimeAttribute(): Carbon
    {
        return static::getCheckinDatetime($this->start_date_str);
    }

    /**
     * @return Carbon|\Carbon\Carbon
     */
    public function getEndDatetimeAttribute(): Carbon
    {
        return static::getCheckoutDatetime($this->end_date_str, $this->checkout)->addDay();
    }

    /**
     * @return bool
     * @SuppressWarnings(PHPMD.BooleanGetMethodName)
     */
    public function getIsPastAttribute(): bool
    {
        return $this->end_datetime->timestamp < static::now();
    }

    /**
     * @return bool
     * @SuppressWarnings(PHPMD.BooleanGetMethodName)
     */
    public function getIsPresentAttribute(): bool
    {
        return $this->start_datetime->timestamp <= static::now() && $this->end_datetime->timestamp >= static::now();
    }

    /**
     * @return bool
     * @SuppressWarnings(PHPMD.BooleanGetMethodName)
     */
    public function getIsFutureAttribute(): bool
    {
        return $this->start_datetime->timestamp > static::now();
    }

    /**
     * @return int
     */
    public function getNightsAttribute(): int
    {
        return $this->end_date->diffInDays($this->start_date) + 1;
    }

    /**
     * @return int
     */
    public function getDaysAttribute(): int
    {
        return $this->nights + 1;
    }

    /**
     * @return string
     */
    public function getStaysAttribute(): string
    {
        return sprintf(__('misc.reservations.stays'), $this->nights, $this->days);
    }

    /**
     * @return int
     */
    public function getChargeAttribute(): int
    {
        return $this->nights * $this->room->price;
    }

    /**
     * @return int
     */
    public function getPaymentAttribute(): int
    {
        return $this->detail->payment ?? 0;
    }

    /**
     * @param  string|null  $startDate
     * @param  string|null  $endDate
     * @param  string|null  $checkoutTime
     *
     * @return bool
     */
    public static function isTermValid(?string $startDate, ?string $endDate, ?string $checkoutTime): bool
    {
        if (empty($startDate) || empty($endDate)) {
            return true;
        }

        if (strtotime($startDate) > strtotime($endDate)) {
            return false;
        }

        $maxDay = Setting::getSetting('max_day');
        if ($maxDay > 0 && $maxDay <= Carbon::parse($endDate)->diffInDays(Carbon::parse($startDate))) {
            return false;
        }

        if ($checkoutTime && static::today()->setTimeFromTimeString($checkoutTime)->isAfter(static::today()->setTimeFromTimeString(self::getCheckinTime()))) {
            return false;
        }

        return true;
    }

    /**
     * @param  int|null  $reservationId
     * @param  int|null  $roomId
     * @param  string|null  $startDate
     * @param  string|null  $endDate
     *
     * @return bool
     */
    public static function isReservationAvailable(?int $reservationId, ?int $roomId, ?string $startDate, ?string $endDate): bool
    {
        if (empty($roomId) || empty($startDate) || empty($endDate)) {
            return true;
        }

        $startDate = self::normalizeDate($startDate);
        $endDate   = self::normalizeDate($endDate);

        $builder = static::where('room_id', $roomId)
                         ->where('end_date', '>=', $startDate)
                         ->where('start_date', '<=', $endDate);
        if (! empty($reservationId)) {
            $builder->where('id', '!=', $reservationId);
        }

        return ! $builder->exists();
    }

    /**
     * @param  int|null  $reservationId
     * @param  int|null  $guestId
     * @param  string|null  $startDate
     * @param  string|null  $endDate
     *
     * @return bool
     */
    public static function isNotDuplicated(?int $reservationId, ?int $guestId, ?string $startDate, ?string $endDate): bool
    {
        if (empty($guestId) || empty($startDate) || empty($endDate)) {
            return true;
        }

        $startDate = self::normalizeDate($startDate);
        $endDate   = self::normalizeDate($endDate);

        $builder = static::where('guest_id', $guestId)
                         ->where('end_date', '>=', $startDate)
                         ->where('start_date', '<=', $endDate);
        if (! empty($reservationId)) {
            $builder->where('id', '!=', $reservationId);
        }

        return ! $builder->exists();
    }

    /**
     * @return bool
     */
    public function checkin()
    {
        $this->status = 'checkin';

        return $this->save();
    }

    /**
     * @param  int|null  $payment
     *
     * @return bool
     */
    public function checkout(?int $payment = null)
    {
        $this->detail->paid($payment);
        $this->status = 'checkout';

        return $this->save();
    }

    /**
     * @return bool
     */
    public function cancel()
    {
        $this->status = 'canceled';

        return $this->save();
    }

    /**
     * @param  string  $date
     *
     * @return string
     */
    // @codeCoverageIgnoreStart
    private static function normalizeDate(string $date)
    {
        if ('sqlite' === config('database.default')) {
            if (strlen($date) <= 10) {
                return $date.' 00:00:00';
            }
        }

        return $date;
    }
    // @codeCoverageIgnoreEnd
}
