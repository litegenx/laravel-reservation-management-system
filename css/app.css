.router-link-active {
  cursor: pointer;
}

