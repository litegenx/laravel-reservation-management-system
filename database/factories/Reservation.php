<?php

use Faker\Generator as Faker;
use App\Models\Reservation;

/** @var Illuminate\Database\Eloquent\Factory $factory */
$factory->define(Reservation::class, function (Faker $faker) {
    $start = $faker->dateTimeBetween('-10days', '+10days')->format('Y-m-d');

    return [
        'start_date' => $start,
        'end_date'   => $faker->dateTimeBetween($start, $start.'  +4 days')->format('Y-m-d'),
        'number'     => $faker->numberBetween(1, 2),
    ];
});
