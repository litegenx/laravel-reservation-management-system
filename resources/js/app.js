require('./bootstrap');
import 'babel-polyfill';
import Vue from 'vue';
import router from './router';
import store from './store';
import App from './components/App';
import vuetify from './plugins/vuetify';

new Vue({
    el: '#app',
    store,
    router,
    components: { App },
    template: '<App/>',
    vuetify,
});
