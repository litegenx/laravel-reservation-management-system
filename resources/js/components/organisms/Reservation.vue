<template>
    <v-dialog
        v-model="dialog"
        max-width="850px"
    >
        <template v-slot:activator="{ on: { click } }">
            <v-btn
                icon
                v-on:click="click"
            >
                <v-icon>mdi-calendar-month</v-icon>
            </v-btn>
        </template>
        <Calendar
            v-if="dialog"
            ref="calendar"
            :event-callback="eventCallback"
            :reset-calendar-callback="resetCalendarCallback"
            :dialog="dialog"
        />
    </v-dialog>
</template>

<script>
    import { mapActions } from 'vuex';
    import moment from 'moment';
    import Calendar from '../molecules/Calendar';

    export default {
        components: {
            Calendar,
        },
        data () {
            return {
                dialog: false,
                format: 'YYYY-MM-DD',
            };
        },
        props: {
            roomId: {
                type: Number,
                required: true,
            },
        },
        methods: {
            ...mapActions({
                search: 'crud/search',
            }),
            async eventCallback (start, end, callback, calendar) {
                const reserves = await this.search({
                    model: 'reservations',
                    query: {
                        'start_date': start.format(this.format),
                        'end_date': end.format(this.format),
                        'room_id': this.roomId,
                    },
                });

                if (calendar) {
                    this.clearAllEvents(calendar);
                }
                callback(reserves.map(this.createEvent));
            },
            createEvent (reserve) {
                const start = moment(reserve.start_datetime);
                const end = moment(reserve.end_datetime);
                return {
                    title: this.getReservationTitle(reserve),
                    start: start.format(this.format),
                    end: end.format(this.format),
                    allDay: true,
                    color: '#0a6',
                    textColor: '#fff',
                    displayEventTime: false,
                    startEditable: false,
                    durationEditable: false,
                    resourceEditable: false,
                };
            },
            getReservationTitle (reserve) {
                return reserve.guest.detail.name;
            },
            resetCalendarCallback (calendar) {
                this.clearAllEvents(calendar);
            },
            clearAllEvents (calendar) {
                this.clearEvent('targetEvent', calendar);
                this.clearEvent('activeEvent', calendar);
            },
            clearEvent (name, calendar) {
                const event = calendar.getEventById(name);
                if (event) {
                    event.remove();
                    this.clearEvent(name, calendar);
                }
            },
        },
    };
</script>
