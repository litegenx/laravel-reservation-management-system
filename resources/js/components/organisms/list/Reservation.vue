<template>
    <v-layout>
        <v-flex>
            <v-list>
                <v-list-item>
                    <v-list-item-content>
                        <v-list-item-title>{{ item.room.name }}</v-list-item-title>
                        <v-list-item-subtitle>部屋名</v-list-item-subtitle>
                    </v-list-item-content>
                </v-list-item>
                <v-list-item>
                    <v-list-item-content>
                        <v-list-item-title>{{ item.room.price }}円</v-list-item-title>
                        <v-list-item-subtitle>金額</v-list-item-subtitle>
                    </v-list-item-content>
                </v-list-item>
                <v-list-item>
                    <v-list-item-content>
                        <v-list-item-title>{{ item.number }}人</v-list-item-title>
                        <v-list-item-subtitle>人数</v-list-item-subtitle>
                    </v-list-item-content>
                </v-list-item>
                <v-list-item>
                    <v-list-item-content>
                        <v-list-item-title>{{ item.start_datetime | moment }}～{{ item.end_datetime | moment }}</v-list-item-title>
                        <v-list-item-subtitle>期間</v-list-item-subtitle>
                    </v-list-item-content>
                </v-list-item>
                <v-list-item>
                    <v-list-item-content>
                        <v-list-item-title>{{ item.guest.detail.name }}（{{ item.guest.detail.name_kana }}）</v-list-item-title>
                        <v-list-item-subtitle>名前</v-list-item-subtitle>
                    </v-list-item-content>
                </v-list-item>
                <v-list-item>
                    <v-list-item-content>
                        <v-list-item-title>{{ item.guest.detail.zip_code }}</v-list-item-title>
                        <v-list-item-subtitle>郵便番号</v-list-item-subtitle>
                    </v-list-item-content>
                </v-list-item>
                <v-list-item>
                    <v-list-item-content>
                        <v-list-item-title>{{ item.guest.detail.address }}</v-list-item-title>
                        <v-list-item-subtitle>住所</v-list-item-subtitle>
                    </v-list-item-content>
                </v-list-item>
                <v-list-item>
                    <v-list-item-content>
                        <v-list-item-title>{{ item.guest.detail.phone }}</v-list-item-title>
                        <v-list-item-subtitle>電話番号</v-list-item-subtitle>
                    </v-list-item-content>
                </v-list-item>
            </v-list>
        </v-flex>
    </v-layout>
</template>

<script>
    import moment from 'moment';

    export default {
        props: ['item'],
        filters: {
            moment: function (date) {
                return moment(date).format('YYYY/MM/DD HH:mm');
            },
        },
    };
</script>
