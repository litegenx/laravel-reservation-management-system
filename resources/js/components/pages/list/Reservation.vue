<template>
    <List
        target-model="reservations"
    />
</template>

<script>
    import List from '../../templates/List';

    export default {
        metaInfo: {
            title: 'Reservation List',
        },
        components: {
            List,
        },
    };
</script>
