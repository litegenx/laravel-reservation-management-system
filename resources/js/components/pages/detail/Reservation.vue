<template>
    <Detail
        target-model="reservations"
        :detail-content-component="Reservation"
    />
</template>

<script>
    import Detail from '../../templates/Detail';
    import Reservation from '../../organisms/detail/Reservation';

    export default {
        data () {
            return {
                Reservation,
            };
        },
        metaInfo: {
            title: 'Reservation Detail',
        },
        components: {
            Detail,
        },
    };
</script>
