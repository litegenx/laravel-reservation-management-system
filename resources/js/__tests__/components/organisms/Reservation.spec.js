import { mount } from '@vue/test-utils';
import Vuex from 'vuex';
import Reservation from '../../../components/organisms/Reservation';
import app from '../app';
import setupLocalVue from '../setupLocalVue';

jest.useFakeTimers();

describe('Reservation', () => {
    it('should render reservation', () => {
        const wrapper = mount(app(Reservation, 'inner', {
            props: {
                roomId: 0,
            },
        }), setupLocalVue({
            store: new Vuex.Store({
                state: {
                    dialog: false,
                    format: 'YYYY-MM-DD',
                },
            }),
        }));
        expect(wrapper.element).toMatchSnapshot();

        expect(wrapper.isVueInstance()).toBeTruthy();
        expect(wrapper.findAll('button')).toHaveLength(1);
    });
});
