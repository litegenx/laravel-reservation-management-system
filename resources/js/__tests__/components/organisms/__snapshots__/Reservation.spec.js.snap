// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reservation should render reservation 1`] = `
<div
  class="v-application v-application--is-ltr theme--light"
  data-app="true"
  id="app"
>
  <div
    class="v-application--wrap"
  >
    <main
      app="true"
      class="v-main"
      classname="mb-5"
      style="padding: 0px 0px 0px 0px;"
    >
      <div
        class="v-main__wrap"
      >
        <div
          class="v-dialog__container"
          role="dialog"
        >
          <button
            class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"
            type="button"
          >
            <span
              class="v-btn__content"
            >
              <i
                aria-hidden="true"
                class="v-icon notranslate mdi mdi-calendar-month theme--light"
              />
            </span>
          </button>
          <!---->
        </div>
        <div
          id="test-contents"
        >
          contents
        </div>
      </div>
    </main>
  </div>
</div>
`;
